ironfish-cli/src/commands/accounts/repair.ts

```diff
@@ -0,0 +1,183 @@
+
+      Assert.isNotUndefined(
+        transactionValue,
+        `Account has a note but is missing the transaction that it received the note from. ${RESCAN_MESSAGE}`,
+      )
+
+      if (!decryptedNoteValue.nullifier) {
+        if (transactionValue.sequence) {
+          throw new Error(
+            `Transaction marked as on chain, but note missing nullifier. ${RESCAN_MESSAGE}`,
+          )
+        }
+
+        continue
+      }
+
+      const spent = await chain.nullifiers.contains(decryptedNoteValue.nullifier)
+
+      if (spent && !decryptedNoteValue.spent) {
+        noteUnspentMismatches++
+
+        await walletDb.saveDecryptedNote(account, decryptedNoteValue.hash, {
+          ...decryptedNoteValue,
+          spent,
+        })
+      } else if (
+        !spent &&
+        !chain.verifier.isExpiredSequence(
+          transactionValue.transaction.expirationSequence(),
+          chain.head.sequence,
+        )
+      ) {
+        unconfirmedBalance += decryptedNoteValue.note.value()
+      }
+    }
+
+    this.log(
+      `\tSaving new unconfirmed balance: ${CurrencyUtils.renderIron(unconfirmedBalance, true)}`,
+    )
+    await walletDb.saveUnconfirmedBalance(account, unconfirmedBalance)
+
+    this.log(
+      `\tRepaired ${noteUnspentMismatches} decrypted notes incorrectly marked as unspent`,
+    )
+  }
+
+  private async repairNullifierToNoteHash(account: Account, walletDb: WalletDB): Promise<void> {
+    let missingNotes = 0
+
+    for await (const [[, nullifier], noteHash] of walletDb.nullifierToNoteHash.getAllIter(
+      undefined,
+      account.prefixRange,
+    )) {
+      const decryptedNoteValue = await account.getDecryptedNote(noteHash)
+
+      if (!decryptedNoteValue) {
+        missingNotes++
+
+        await walletDb.deleteNullifier(account, nullifier)
+      }
+    }
+
+    this.log(
+      `\tRepaired ${missingNotes} nullifiers that map to notes that are not in the wallet`,
+    )
+  }
+
+  private async repairSequenceToNoteHash(account: Account, walletDb: WalletDB): Promise<void> {
+    let incorrectSequences = 0
+
+    for await (const [, [sequence, noteHash]] of walletDb.sequenceToNoteHash.getAllKeysIter(
+      undefined,
+      account.prefixRange,
+    )) {
+      const decryptedNoteValue = await account.getDecryptedNote(noteHash)
+
+      if (!decryptedNoteValue) {
+        incorrectSequences++
+
+        await walletDb.sequenceToNoteHash.del([account.prefix, [sequence, noteHash]])
+
+        continue
+      }
+
+      const transactionValue = await account.getTransaction(decryptedNoteValue.transactionHash)
+
+      Assert.isNotUndefined(
+        transactionValue,
+        `Account has a note but is missing the transaction that it received the note from. ${RESCAN_MESSAGE}`,
+      )
+
+      if (transactionValue.sequence !== sequence) {
+        incorrectSequences++
+
+        await walletDb.sequenceToNoteHash.del([account.prefix, [sequence, noteHash]])
+      }
```

Comment:
should this put it at the correct sequence?

Reply:
Updated